pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/stats/MapColumnPreIndexStatsCollector.java

```diff
@@ -0,0 +1,288 @@
+            updatePartition(key);
+          }
+          if (_minValue == null) {
+            _minValue = key;
+          } else {
+            if (key.compareTo(_minValue) < 0) {
+              _minValue = key;
+            }
+          }
+          if (_maxValue == null) {
+            _maxValue = key;
+          } else {
+            if (key.compareTo(_maxValue) > 0) {
+              _maxValue = key;
+            }
+          }
+          _minLength = Math.min(_minLength, length);
+          _maxLength = Math.max(_maxLength, length);
+        }
+
+        // Record statistics about the value within the key
+        AbstractColumnStatisticsCollector keyStats = getOrCreateKeyStatsCollector(key, mapValueEntry.getValue());
+        keyStats.collect(mapValueEntry.getValue());
+      }
+      _totalNumberOfEntries++;
+    } else {
+      throw new UnsupportedOperationException();
+    }
+  }
+
+  @Override
+  public String getMinValue() {
+    if (_sealed) {
+      return _minValue;
+    }
+    throw new IllegalStateException("you must seal the collector first before asking for min value");
+  }
+
+  @Override
+  public String getMaxValue() {
+    if (_sealed) {
+      return _maxValue;
+    }
+    throw new IllegalStateException("you must seal the collector first before asking for max value");
+  }
+
+  @Override
+  public String[] getUniqueValuesSet() {
+    if (_sealed) {
+      return _sortedValues;
+    }
+    throw new IllegalStateException("you must seal the collector first before asking for unique values set");
+  }
+
+  @Override
+  public int getLengthOfShortestElement() {
+    return _minLength;
+  }
+
+  @Override
+  public int getLengthOfLargestElement() {
+    if (_sealed) {
+      return _maxLength;
+    }
+    throw new IllegalStateException("you must seal the collector first before asking for longest value");
+  }
+
+  @Override
+  public int getMaxRowLengthInBytes() {
+    return _maxRowLength;
+  }
+
+  @Override
+  public int getCardinality() {
+    if (_sealed) {
+      return _sortedValues.length;
+    }
+    throw new IllegalStateException("you must seal the collector first before asking for cardinality");
+  }
+
+  @Override
+  public void seal() {
+    if (!_sealed) {
+      _sortedValues = _keys.stream().sorted().toArray(String[]::new);
+      _keys = null;
+
+      // Iterate through every key stats collector and seal them
+      for (AbstractColumnStatisticsCollector keyStatsCollector : _keyStats.values()) {
+        keyStatsCollector.seal();
+      }
+
+      _sealed = true;
+    }
+  }
+
+  /**
+   * Create a Stats Collector for each Key in the Map.
+   *
+   * NOTE: this could raise an issue if there are millions of keys with very few values (Sparse keys, in other words).
+   * So a less memory intensive option may be better for this.
+   *
+   * @param key
+   * @param value
+   * @return
+   */
+  private AbstractColumnStatisticsCollector getOrCreateKeyStatsCollector(String key, Object value) {
```

Comment:
How expensive is this Stats Collector specially if large number of keys are there in MAP?

Reply:
It will be expensive, we could optimize more later.